pkg/logql/accumulator.go

```diff
@@ -0,0 +1,379 @@
+package logql
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/grafana/loki/pkg/logproto"
+	"github.com/grafana/loki/pkg/logqlmodel"
+	"github.com/grafana/loki/pkg/logqlmodel/metadata"
+	"github.com/grafana/loki/pkg/logqlmodel/stats"
+	"github.com/grafana/loki/pkg/querier/queryrange/queryrangebase/definitions"
+	"github.com/grafana/loki/pkg/util/math"
+)
+
+// NewBufferedAccumulator returns an accumulator which aggregates all query
+// results in a slice. This is useful for metric queries, which are generally
+// small payloads and the memory overhead for buffering is negligible.
+func NewBufferedAccumulator(n int) *BufferedAccumulator {
+	return &BufferedAccumulator{
+		results: make([]logqlmodel.Result, n),
+	}
+}
+
+type BufferedAccumulator struct {
+	results []logqlmodel.Result
+}
+
+func (a *BufferedAccumulator) Accumulate(_ context.Context, acc logqlmodel.Result, i int) error {
+	a.results[i] = acc
+	return nil
+}
+
+func (a *BufferedAccumulator) Result() []logqlmodel.Result {
+	return a.results
+}
+
+type quantileSketchAccumulator struct {
```

Comment:
Should we make all accumulator implementations exported?

```suggestion
type QuantileSketchAccumulator struct {
```